data/blog/boosting-mobile-session-replay-performance-with-view-renderer-v2.md

```diff
@@ -0,0 +1,312 @@
+
+To correct the timing the system **skips** the frames that should already have been rendered at this point in time (the _dropped frames)_. This correction behavior is commonly known as _frame drops_ and has the undesired side effect of intermittent frame rate changes that can be noticed as visual stutters (i.e. not scrolling smoothly).
+
+<div align="center">
+  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/frame-drops.jpg" alt="Dropped Frames"/>
+  _When workload takes longer than available time, we need to drop frames_
+</div>
+
+Frame drops were also the unintended side effect for the previous implementation of Session Replay, as the once-per-second screenshot simply took too long.
+
+# Taking Screenshots With PII In Mind
+
+Now that we know what the issue is, we need to take a closer look at the work done by Session Replay. The capture process consists of three steps that must be executed for each screenshot:
+
+1. **Redact**: Traverse the view hierarchy to identify UI elements that contain sensitive information and require masking.
+2. **Render**: Convert the current view hierarchy into a rasterized bitmap representation ("take a screenshot") using a _view renderer_.
+3. **Mask**: Apply privacy masking by combining the rendered image with the redaction geometry.
+
+As mentioned before, accessing the view hierarchy must be done on the main thread, which means that both steps _Redact_ and _Render_ must also execute on the main thread. Only the final step _Mask_ can be performed on a background queue, since it operates solely on the image data produced by the previous steps.
+
+This allows us to focus on the performance of the _Redact_ and _Render_ steps.
+
+<div align="center">
+  <img src="/images/boosting-mobile-session-replay-performance-with-view-renderer-v2/uikit-meme.png" alt="UIKit Meme"/>
+  _Source: [@ios_memes on X](https://x.com/ios_memes/status/1483418141871030276)_
+</div>
+
+# Measuring Baseline Performance
+
+After adding some code to calculate the execution time and adding a basic sampling mechanism to reduce variation over time, we quickly gathered our first insights using a (rather old) iPhone 8:
+
+**The step _Render_ takes `~155ms` per frame, causing ~9-10 of 60 frames being dropped every second of using Session Replay** 💀
+
+| 120 samples | Redact        | Render          | Total           |
+| ----------- | ------------- | --------------- | --------------- |
+| Min         | 3.0583 ms     | 145.3815 ms     | 151.4525 ms     |
+| Avg         | 5.8453 ms     | 149.8243 ms     | 155.6732 ms     |
+| p50         | 6.0484 ms     | 149.2103 ms     | 154.1397 ms     |
+| p75         | 6.1136 ms     | 151.9487 ms     | 158.0255 ms     |
+| p95         | **6.2567 ms** | **155.3496 ms** | **161.3549 ms** |
```

Comment:
Is it possible to make these measurements a different colour? The bolding is so subtle, I almost missed it :o

Reply:
Changed it to be bold and in the brand color.